packages/ai/core/tool/mcp/client.ts

```diff
@@ -0,0 +1,302 @@
+import { z, ZodType } from 'zod';
+import {
+  CallToolRequest,
+  CallToolResult,
+  CallToolResultSchema,
+  Implementation,
+  InitializeResultSchema,
+  JSONRPCError,
+  JSONRPCNotification,
+  JSONRPCRequest,
+  JSONRPCResponse,
+  LATEST_PROTOCOL_VERSION,
+  ListToolsRequest,
+  ListToolsResult,
+  ListToolsResultSchema,
+  Notification,
+  Request,
+  RequestOptions,
+  SUPPORTED_PROTOCOL_VERSIONS,
+  Transport,
+  TransportConfig,
+} from './types';
+import { AISDKError } from '@ai-sdk/provider';
+import { createMcpTransport } from './transport';
+
+const CONNECTION_TIMEOUT_MS = 6000;
+const REQUEST_TIMEOUT_MS = 3000;
+
+interface MCPClientConfig {
+  transport: TransportConfig;
+  name?: string;
+  version?: string;
+  connectionTimeoutMs?: number;
+  requestTimeoutMs?: number;
+}
+
+/**
+ * A lightweight MCP Client implementation,
+ * primarily for tool conversion between MCP<>AI SDK.
+ *
+ * It is a custom implementation of the MCP Client (derived from Protocol) class.
+ *
+ * Not supported:
+ * - Client options (e.g. sampling, roots) as they are not needed for tool conversion
+ * - Accepting notifications
+ */
+export class MCPClient {
+  private transport: Transport;
+  private clientInfo: Implementation;
+  private requestMessageId = 0;
+  private responseHandlers: Map<
+    number,
+    (response: JSONRPCResponse | Error) => void
+  > = new Map();
+  private connectionTimeoutMs: number;
+  private requestTimeoutMs: number;
+
+  constructor({
+    transport: transportConfig,
+    name = 'ai-sdk-mcp-client',
+    version = '1.0.0',
+    connectionTimeoutMs = CONNECTION_TIMEOUT_MS,
+    requestTimeoutMs = REQUEST_TIMEOUT_MS,
+  }: MCPClientConfig) {
+    this.transport = createMcpTransport(transportConfig);
+    this._initTransport();
+    this.clientInfo = {
+      name,
+      version,
+    };
+    this.connectionTimeoutMs = connectionTimeoutMs;
+    this.requestTimeoutMs = requestTimeoutMs;
+  }
+
+  async init(): Promise<this> {
```

Comment:
why is there an init that's separate from the constructor and `initTransport`?

Reply:
ah that was in response to your comment regarding merging the construction + initialization of the connection! since constructor cannot have async calls!